### PR TITLE
Stop the bork - gracefully handle server errors (Error PR 1 of 3)

### LINKED ISSUE #1774: Use raven to handle server errors
### Background
Twice now we've had a white screen on production for certain users because notifications were borked. If a certain component is borked, it shouldn't degrade the ENTIRE experience, just that piece.

### Acceptance Criteria (optional)
- [ ] Establish a best practice on where to put error boundaries 
- [ ] Put error boundaries where they should go so there is minimal experience degradation
- [ ] call raven inside the error boundary so we still know what went wrong

- **Estimated effort:** 12 points ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))

### PR DESCRIPTION
**Pull request by mattkrick**
Recently, we've been fighting too many fires.
We goof up, and instead of that goof up gracefully landing in sentry, the customer's screen goes white & there are no errors in the console.
This fixes that. Fixes #1774

There were a few things going wrong:
- The dataloader would throw an error if there was a missing foreign key. Albeit if a foreign key doesn't exist, it is 100% our fault, but the penalty for failure was too severe (blank screen of death). Now, we log it and send down the null value to the client. It's then up to the client to handle it gracefully
- The client was not handling errors gracefully for HTTP requests. This was my fault. I now pass whatever values, good or bad, into relay & then we can handle it with onError callbacks if need be.
- The client was not handling errors gracefully for websocket requests. This was Apollo's fault. When GraphQL encounters an error, it sends the data that it _could_ fetch, plus the error that it encountered. _Apollo ignores all that beautiful 99% correct data_. I can't fix this without rewriting the client, so for now, I trick it by having the server tell it it's not an error. 
- We weren't sending any errors to sentry from the server. Now we do, so we no longer have to `dokku logs web`. yay! 

TEST
- [ ] Create 2 users, as user A, assign a task to user B so they get a "TASK_INVOLVES" notification, exit the browser for user A.
- [ ] go into the database (localhost:8080) find that notification & change the type to "X"
- [ ] Refresh the page, notice it doesn't white screen.
- [ ] look at the server log (locally), notice it shows an error
- [ ] run `yarn bs` & use the SENTRY_DNS key that's on the staging server environment, or just run this on the dev server. Repeat the steps. Instead of seeing an error in the server console, you see nothing, but an error does show up in sentry (example: https://sentry.io/parabol/action-staging/issues/484351180/)
- [ ] go into the database & revert the type back to TASK_INVOLVES but now set the notification's taskId to 'X'.
- [ ] Refresh & notice there isn't a blank white page
- [ ] Get a similar message sent to sentry (example: https://sentry.io/parabol/action-staging/issues/484298502/events/15455592846/)

### CODE REVIEW

**Comment by jordanh** [comment]
# Test Cycle 1
## Status: in progress
- [ ] Create 2 users, as user A, assign a task to user B so they get a "TASK_INVOLVES" notification, exit the browser for user A.
- [ ] go into the database (localhost:8080) find that notification & change the type to "X"
- [ ] Refresh the page, notice it doesn't white screen.
- [ ] look at the server log (locally), notice it shows an error
- [ ] run `yarn bs` & use the SENTRY_DNS key that's on the staging server environment, or just run this on the dev server. Repeat the steps. Instead of seeing an error in the server console, you see nothing, but an error does show up in sentry (example: https://sentry.io/parabol/action-staging/issues/484351180/)
- [ ] go into the database & revert the type back to TASK_INVOLVES but now set the notification's taskId to 'X'.
- [ ] Refresh & notice there isn't a blank white page
- [ ] Get a similar message sent to sentry (example: https://sentry.io/parabol/action-staging/issues/484298502/events/15455592846/)

File: src/server/server.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ import cors from 'cors';
 import bodyParser from 'body-parser';
 import jwt from 'express-jwt';
 import favicon from 'serve-favicon';
-import raven from 'raven';
+import Raven from 'raven';
```
Comment (jordanh) [inline src/server/server.js]: Wait. It was a capitalization problem?? Ugh.

File: src/server/server.js [hunk 2]
```diff
@@ -58,10 +61,14 @@ if (!PROD) {
   app.use(require('webpack-hot-middleware')(compiler)); // eslint-disable-line import/no-extraneous-dependencies
   /* eslint-enable global-require */
 }
-
+Raven.config(process.env.SENTRY_DSN, {
+  release: version,
+  environment: process.env.NODE_ENV,
+  parseUser: jwtFields
```
Comment (jordanh) [inline src/server/server.js]: Awesome.

File: src/server/server.js [hunk 3]
```diff
@@ -58,10 +61,14 @@ if (!PROD) {
...
+}).install();
 // setup middleware
 // sentry.io request handler capture middleware, must be first:
-app.use(raven.requestHandler(process.env.SENTRY_DSN));
+app.use(Raven.requestHandler());
```
Comment (jordanh) [inline src/server/server.js]: Do we need to conditionally not use this middleware if SENTRY_DSN is not set to re-enable server console logs?

Comment (mattkrick) [reply src/server/server.js]: server logs are based off the the NODE_ENV, not the presence of sentry

**Comment by jordanh** [comment]
Ok, first cycle complete! Have a look at my notes comments. Penny for your thoughts. I think we're quite close to a merge.

File: src/server/server.js (hunk 1 above)
Comment (mattkrick) [reply src/server/server.js]: nawh, we bumped up a version & this is how they did it in the docs. I'm assuming the capitalize because it's a singleton? i dunno, ES2015 has straight up bastardized the when-to-capitalize rules

File: src/server/server.js (hunk 3 above)
Comment (jordanh) [reply src/server/server.js]: Hmm, that isn't the behavior I was observing when I ran my test using `yarn dev`.

**Comment by mattkrick** [comment]
We'll still need to make some other changes before we merge (I'll make a separate PR for it).

tl;dr not treating things as errors swings too far in the other direction. we're trading no-white-screen for no-server-validation-error-messages.


So the big problem with error handling in graphql is that you can have an error AND data.
On the client, we need to decide whether it is a success (onCompleted) or a failure (onError).
If we call it a failure, we don't get the data & that means we get the white screen of death.
So, we have 2 options:
- call everything a success
- base the decision off of analyzing the payload. for example, maybe `{foo: null}` is a failure. But what if the field is deep in a schema? for example, `{foo: {bar: null}}` gives us no real info, so we call that a failure, too? what if we start batching queries & 1 fails & the other succeeds?

Given those 2 options, i say we call every graphql response a success because otherwise the rules are just arbitrary. Therefore, the only "error" must occur before graphql got called (eg cannot connect to the server, our server rate-limited or blocked the ip, invalid auth, etc).

So, given that the Network layer treats every graphql response as a success, we now need to figure out how to handle the response errors.

For mutations, that's a pretty easy refactor job. We just call `onCompleted(res, errors)`. Then, we just move the existing onError callback inside the onCompleted callback.

For queries, we want to use the errors to do thing like propagate user-facing error messages.
For that case, we have 2 options:
- on the server, but these errors into the schema itself & send them down as a field of the schema (like we saw at graphql summit & what we do for accepting email invitations)
- grab the errors in the network layer & shove them into the data object. this will require extending the query with an errors field to make sure relay doesn't sanitize it away.

So, the cleaner option of the 2 is to put errors in the schema. this will be more work, but honestly it solves a tension of mine which was, "what if rethinkdb throws one of those huge errors & the client stuffs it into an error field", This happens in development!

Finally, for subscriptions, it's a bit harder. onCompleted refers to a subscription ending (eg unsubscribe or kickout), so instead we use onNext or onError.

How we handle it on the client depends on how we handle graphql errors on the server. So, let's look at when we throw a gql error:
1. When the client sends us query without a query
2. When a mutation is run using the websocket transport & the result includes an errors prop next to data
3. when a subscription is mal-formed
4. when a subscription throws inside the subscribe resolver (eg is not authorized)
5. when the subscription payload is put together & there is an error (eg a foreign key lookup failed)

How i think we should handle them:
1. Error
2. Not an error
3. Error
4. Error, but doesn't really matter
5. Not an error

So, on the errors, we send down a GQL_ERROR event, else we send down GQL_DATA.
Then, on the client we base what observer method to call based on the event type.

...and that's it! All in all, it makes a whole lot more sense than what we're doing now. Forcing client-facing error messages to be part of the schema is not our norm, but it's healthier because we can break things out by eg message, category, subcategory, language, etc. Before, we accomplished this by parsing the error message, which wasn't great.

File: src/server/server.js (hunk 3 above)
Comment (mattkrick) [reply src/server/server.js]: ah, i think i know what you were talking about, where sentry still sends a warning to the console? i fixed that in #1869

**Comment by mattkrick** [comment]
UPDATE:
I broke this into a 3 part PR for easier review (part 2 is still huge though). They're a package deal-- if 1 goes to production, they all should go

**Comment by jordanh** [comment]
# Test Cycle 2
## Status: in progress
- [ ] Create 2 users, as user A, assign a task to user B so they get a "TASK_INVOLVES" notification, exit the browser for user A.
- [ ] go into the database (localhost:8080) find that notification & change the type to "X"
- [ ] Refresh the page, notice it doesn't white screen.
- [ ] look at the server log (locally), notice it shows an error
- [ ] run `yarn bs` & use the SENTRY_DNS key that's on the staging server environment, or just run this on the dev server. Repeat the steps. Instead of seeing an error in the server console, you see nothing, but an error does show up in sentry (example: https://sentry.io/parabol/action-staging/issues/484351180/)
- [ ] go into the database & revert the type back to TASK_INVOLVES but now set the notification's taskId to 'X'.
- [ ] Refresh & notice there isn't a blank white page
- [ ] Get a similar message sent to sentry (example: https://sentry.io/parabol/action-staging/issues/484298502/events/15455592846/)

## Notes

**Comment by jordanh** [comment]
@mattkrick here are the results from test cycle 2. We should probably make sure errors emit on the conole when we run `npm run dev` / `yarn dev`. What do you think?

**Comment by mattkrick** [comment]
@jordanh printing to server console in dev is fixed in #1869. gonna go ahead & merge em both
